core/src/main/java/org/elasticsearch/action/support/single/custom/TransportSingleCustomOperationAction.java

```diff
@@ -123,7 +124,32 @@ private AsyncSingleAction(Request request, ActionListener<Response> listener) {
             if (blockException != null) {
                 throw blockException;
             }
-            this.shardsIt = shards(clusterState, internalRequest);
+            ShardsIterator originalShardsIt = shards(clusterState, internalRequest);
```

Comment:
typically this kind of logic (preferring local shards) is typically folded into the shard iterator. Why do we need a custom implementation here? I know it was there before, but if we're cleaning up, it's a better clean up no?